OpenRA.Mods.Common/Traits/World/Pathfinder/BasePathSearch.cs

```diff
@@ -0,0 +1,172 @@
+		IEnumerable<Pair<CPos, int>> Considered { get; }
+		bool Debug { get; set; }
+		Player Owner { get; }
+		int MaxCost { get; }
+		IPathSearch Reverse();
+		IPathSearch WithCustomBlocker(Func<CPos, bool> customBlock);
+		IPathSearch WithIgnoredActor(Actor b);
+		IPathSearch WithHeuristic(Func<CPos, int> h);
+		IPathSearch WithCustomCost(Func<CPos, int> w);
+		IPathSearch WithoutLaneBias();
+		IPathSearch FromPoint(CPos from);
+		bool IsTarget(CPos location);
+		CPos Expand();
+	}
+
+	public abstract class BasePathSearch : IPathSearch
+	{
+		// The Id of a Pathsearch is computed by its properties.
+		// So two PathSearch instances with the same parameters will
+		// Compute the same Id. This is used for caching purposes.
+		public string Id
+		{
+			get
+			{
+				if (string.IsNullOrEmpty(id))
+				{
+					var builder = new StringBuilder();
+					builder.Append(this.Actor.ActorID);
+					while (!startPoints.Empty)
+					{
+						var startpoint = startPoints.Pop();
+						builder.Append(startpoint.X);
+						builder.Append(startpoint.Y);
+						builder.Append(CellInfo[startpoint].EstTotal);
+					}
+
+					builder.Append(inReverse);
+					if (ignoredActor != null) builder.Append(ignoredActor.ActorID);
+					builder.Append(laneBias);
+
+					id = builder.ToString();
+				}
+
+				return id;
+			}
+		}
+
+		public abstract CellLayer<CellInfo> CellInfo { get; protected set; }
+		public IPriorityQueue<CPos> OpenQueue { get; protected set; }
+		public abstract IEnumerable<Pair<CPos, int>> Considered { get; }
+
+		public Player Owner { get { return this.Actor.Owner; } }
+		public int MaxCost { get; protected set; }
+		public bool Debug { get; set; }
+
+		protected readonly IMobileInfo MobileInfo;
+		protected readonly IActor Actor;
+		protected readonly IWorld World;
+		string id;
+		protected Func<CPos, int> heuristic;
+		protected bool checkForBlocked;
+		protected IActor ignoredActor;
+		protected bool inReverse;
+		protected Func<CPos, int> customCost;
+		protected Func<CPos, bool> customBlock;
+		protected int laneBias = 1;
+
+		// This member is used to compute the ID of PathSearch.
+		// Essentially, it represents a collection of the initial
+		// points considered and their Heuristics to reach
+		// the target. It pretty match identifies, in conjunction of the Actor,
+		// a deterministic set of calculations
+		protected IPriorityQueue<CPos> startPoints;
+
+		protected BasePathSearch(IWorld world, IMobileInfo mobileInfo, IActor actor)
+		{
+			this.Actor = actor;
+			this.World = world;
+			this.MobileInfo = mobileInfo;
+			Debug = false;
+			customCost = null;
+			MaxCost = 0;
+		}
+
+		protected static Func<CPos, int> DefaultEstimator(CPos destination)
+		{
+			return here =>
+			{
+				var diag = Math.Min(Math.Abs(here.X - destination.X), Math.Abs(here.Y - destination.Y));
+				var straight = Math.Abs(here.X - destination.X) + Math.Abs(here.Y - destination.Y);
+
+				// HACK: this relies on fp and cell-size assumptions.
```

Comment:
So, what do you think is the point of that 1.001? Why D = 100 and not another number?